### PR TITLE
Fix vanilla components not being translated

### PR DESCRIPTION
**Pull request by booky10**
Vanilla components were just converted to normal json strings without being translated by the `GlobalTranslator`. This caused custom name's of entities to no longer be translated properly after a server restart or an entity unload.

This does have a small performance cost, as all vanilla components (not backed by an adventure component) are now converted to an adventure component by serializing to json and deserializing from json again.

<details>
<summary>Test plugin diff</summary>

```diff
Index: test-plugin/src/main/java/io/papermc/testplugin/TestPlugin.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/test-plugin/src/main/java/io/papermc/testplugin/TestPlugin.java b/test-plugin/src/main/java/io/papermc/testplugin/TestPlugin.java
--- a/test-plugin/src/main/java/io/papermc/testplugin/TestPlugin.java	(revision c71e79c2d56d2cd374995cbd5a5ac7c3d1fcf5b2)
+++ b/test-plugin/src/main/java/io/papermc/testplugin/TestPlugin.java	(date 1698715701243)
@@ -1,5 +1,16 @@
 package io.papermc.testplugin;
 
+import io.papermc.paper.event.player.AsyncChatEvent;
+import java.text.MessageFormat;
+import java.util.Locale;
+import net.kyori.adventure.text.Component;
+import net.kyori.adventure.translation.GlobalTranslator;
+import net.kyori.adventure.translation.TranslationRegistry;
+import org.bukkit.Bukkit;
+import org.bukkit.Location;
+import org.bukkit.NamespacedKey;
+import org.bukkit.entity.ArmorStand;
+import org.bukkit.event.EventHandler;
 import org.bukkit.event.Listener;
 import org.bukkit.plugin.java.JavaPlugin;
 
@@ -8,5 +19,20 @@
     @Override
     public void onEnable() {
         this.getServer().getPluginManager().registerEvents(this, this);
+
+        TranslationRegistry registry = TranslationRegistry.create(new NamespacedKey(this, "i18n"));
+        registry.register("testplugin.armorstand", Locale.ENGLISH, new MessageFormat("Success!"));
+        GlobalTranslator.translator().addSource(registry);
+    }
+
+    @EventHandler
+    public void onChat(AsyncChatEvent event) {
+        Bukkit.getScheduler().runTask(this, () -> {
+            Location spawnLoc = event.getPlayer().getLocation();
+            spawnLoc.getWorld().spawn(spawnLoc, ArmorStand.class, armorStand -> {
+                armorStand.setCustomNameVisible(true);
+                armorStand.customName(Component.translatable("testplugin.armorstand"));
+            });
+        });
     }
 }

```
</details>

### CODE REVIEW

**Review by electronicboy — APPROVED** [review]

**Comment by Machine-Maker** [comment]
I wonder if it's worth hooking into vanilla's serialization logic (gson in this version, codecs in the next version) and render the translatable components when serializing to json. That would avoid the double conversion.

**Comment by booky10** [comment]
How would that work with adventure's translator? The adventure translator only accepts adventure components, not vanilla components or json.

**Comment by lynxplay** [comment]
Well technically there is Translator's `MessageFormat translate(final @NotNull String key, final @NotNull Locale locale);`
that could be used, however that was recently expanded to also have a method that takes a TranslatableComponent.

**Comment by Machine-Maker** [comment]
I was thinking along the lines of... if, while serializing a vanilla component, it encounters a translatable, it can only convert that component to adventure, render it, and then serialize it to json. That way, *most* components, or chunks of components avoid the double serialization

**Comment by booky10** [comment]
I think I've now implemented your suggested solution to this problem. I've used a `ThreadLocal` to store the adventure locale, as the static gson serializer doesn't contain (or support) any context as far as I know. I'm open to suggestions on alternatives to store the locale.

**Comment by Machine-Maker** [comment]
I think it can be even better. Can't we first check if the translation key exists in adventure's global translator? If it doesn't, do we need to bother running it through adventure's renderer? Maybe @kashike can weigh in on this if I'm missing some functionality of adventure's translation system.

**Comment by booky10** [comment]
> I think it can be even better. Can't we first check if the translation key exists in adventure's global translator? If it doesn't, do we need to bother running it through adventure's renderer? Maybe @kashike can weigh in on this if I'm missing some functionality of adventure's translation system.

I'm not sure this is possible. The adventure `Translator` interface accepts `TranslatableComponent`s or translation keys, but doesn't provide API for listing registered translation keys. A `TranslationRegistry` does have this information.

**Comment by Machine-Maker** [comment]
Hmm yeah... I guess there isn't a way to know if it can be converted. 

This might be as good as it gets then for now.

File: patches/server/0010-Adventure.patch
Comment (dawon) [inline patches/server/0010-Adventure.patch]: You can use instanceof cast, that would simplify the code and make it more readable:

```java
if (component instanceof AdventureComponent aComp) {
           // don't convert to json and back if already backed by an adventure Component
           return aComp.adventure;
}
```

Comment (booky10) [reply patches/server/0010-Adventure.patch]: This style is not used anywhere else in the class (see line 1030), Idk if this should be changed.

Comment (Machine-Maker) [reply patches/server/0010-Adventure.patch]: Well... some of the class was written before that was a thing. I think we'd prefer to use the newer instanceof where we can.

Comment (booky10) [reply patches/server/0010-Adventure.patch]: Ok, changed :+1:

**Comment by booky10** [comment]
I've rewritten this PR for 1.20.3/1.20.4. This now translates translatable components when encoding using the serialization codec. I don't have much experience in working with Codecs, so I don't know if there is a more optiomal spot for converting and translating the components.

**Review by Machine-Maker — DISMISSED** [review]
So while I think the changes work... I would rather not use a ThreadLocal for this I think. Since this CODEC is global, it runs the risk of causing a problem. Lets say in someone's custom Translator, they do something that causes that CODEC to be used, since its on the same thread it will cause the rendering to take effect. I would rather there be a separate "rendering" codec that can be created with a locale. Could be a global map of them with the locale as the key, created as needed. Then that codec would be used to convert the component with the rendering/translating.

This would be done by extracting out that last return statement in createCodec into a separate method and returning the direct Codec<Component> into a field instead. Then, a separate method could be created that took a Locale and wrapped that component in the eithers with the rendering logic.

**Comment by booky10** [comment]
I've now removed the ThreadLocal now and added a map with the translating codecs. I'm not quite sure how storing the direct codec would work, as I think children of components wouldn't be translated because its a recursive codec and we would only translate the component if the parent is also translatable.

**Comment by Machine-Maker** [comment]
What you ended up doing is fine, it is probably just possible to not have to recreate the whole codec each time, and still retain the recursive nature of it. But idk how big a deal that is, not like its gonna be creating a zillion codecs anyways.

**Comment by booky10** [comment]
<details>
<summary>ComponentSerialization Diff</summary>

``` patch
Index: src/main/java/net/minecraft/network/chat/ComponentSerialization.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/net/minecraft/network/chat/ComponentSerialization.java b/src/main/java/net/minecraft/network/chat/ComponentSerialization.java
--- a/src/main/java/net/minecraft/network/chat/ComponentSerialization.java	(revision 553d84e32f0649d97ac888f9a96c3f5d1d954c2c)
+++ b/src/main/java/net/minecraft/network/chat/ComponentSerialization.java	(date 1702305732428)
@@ -32,6 +32,7 @@
     });
 
     // Paper start - adventure localization
+    private static final Codec<Component> DIRECT_CODEC = ExtraCodecs.recursive("Component", ComponentSerialization::createDirectCodec);
     private static final java.util.Map<java.util.Locale, Codec<Component>> TRANSLATING_CODECS = new java.util.HashMap<>();
 
     public static Codec<Component> getTranslatingCodec(@javax.annotation.Nullable java.util.Locale locale) {
@@ -69,18 +70,23 @@
         return ExtraCodecs.orCompressed(mapCodec3, mapCodec2);
     }
 
-    private static Codec<Component> createCodec(Codec<Component> selfCodec) {
-        // Paper start - adventure localization
-        return createCodec(selfCodec, null);
-    }
-
-    private static Codec<Component> createCodec(Codec<Component> selfCodec, @javax.annotation.Nullable java.util.Locale locale) {
-        // Paper end
+    private static Codec<Component> createDirectCodec(Codec<Component> selfCodec) { // Paper - adventure localization
         ComponentContents.Type<?>[] types = new ComponentContents.Type[]{PlainTextContents.TYPE, TranslatableContents.TYPE, KeybindContents.TYPE, ScoreContents.TYPE, SelectorContents.TYPE, NbtContents.TYPE};
         MapCodec<ComponentContents> mapCodec = createLegacyComponentMatcher(types, ComponentContents.Type::codec, ComponentContents::type, "type");
         Codec<Component> codec = RecordCodecBuilder.create((instance) -> {
             return instance.group(mapCodec.forGetter(Component::getContents), ExtraCodecs.strictOptionalField(ExtraCodecs.nonEmptyList(selfCodec.listOf()), "extra", List.of()).forGetter(Component::getSiblings), Style.Serializer.MAP_CODEC.forGetter(Component::getStyle)).apply(instance, MutableComponent::new);
         });
+        // Paper start - adventure localization
+        return codec;
+    }
+
+    private static Codec<Component> createCodec(Codec<Component> selfCodec) {
+        return createCodec(selfCodec, null);
+    }
+
+    private static Codec<Component> createCodec(Codec<Component> selfCodec, final @javax.annotation.Nullable java.util.Locale locale) {
+        final Codec<Component> codec = DIRECT_CODEC;
+        // Paper end
         return Codec.either(Codec.either(Codec.STRING, ExtraCodecs.nonEmptyList(selfCodec.listOf())), codec).xmap((either) -> {
             return either.map((either2) -> {
                 return either2.map(Component::literal, ComponentSerialization::createFromList);
```

</details>

<details>
<summary>TestPlugin Diff</summary>

```diff
Index: test-plugin/src/main/java/io/papermc/testplugin/TestPlugin.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/test-plugin/src/main/java/io/papermc/testplugin/TestPlugin.java b/test-plugin/src/main/java/io/papermc/testplugin/TestPlugin.java
--- a/test-plugin/src/main/java/io/papermc/testplugin/TestPlugin.java	(revision 75d74272d91cc1b00a2e552d61b20a1d57cae799)
+++ b/test-plugin/src/main/java/io/papermc/testplugin/TestPlugin.java	(date 1702305590501)
@@ -1,5 +1,16 @@
 package io.papermc.testplugin;
 
+import io.papermc.paper.event.player.AsyncChatEvent;
+import java.text.MessageFormat;
+import java.util.Locale;
+import net.kyori.adventure.text.Component;
+import net.kyori.adventure.translation.GlobalTranslator;
+import net.kyori.adventure.translation.TranslationRegistry;
+import org.bukkit.Bukkit;
+import org.bukkit.Location;
+import org.bukkit.NamespacedKey;
+import org.bukkit.entity.ArmorStand;
+import org.bukkit.event.EventHandler;
 import org.bukkit.event.Listener;
 import org.bukkit.plugin.java.JavaPlugin;
 
@@ -8,5 +19,28 @@
     @Override
     public void onEnable() {
         this.getServer().getPluginManager().registerEvents(this, this);
+
+        TranslationRegistry registry = TranslationRegistry.create(new NamespacedKey(this, "i18n"));
+        registry.register("testplugin.armorstand", Locale.ENGLISH, new MessageFormat("Success! ({0})"));
+        registry.register("testplugin.armorstand.arg", Locale.ENGLISH, new MessageFormat("Argument"));
+        GlobalTranslator.translator().addSource(registry);
+    }
+
+    @EventHandler
+    public void onChat(AsyncChatEvent event) {
+        Bukkit.getScheduler().runTask(this, () -> {
+            Location spawnLoc = event.getPlayer().getLocation();
+            spawnLoc.getWorld().spawn(spawnLoc, ArmorStand.class, armorStand -> {
+                armorStand.setCustomNameVisible(true);
+                armorStand.customName(Component.translatable("testplugin.armorstand",
+                    Component.translatable("testplugin.armorstand.arg")));
+            });
+            spawnLoc.add(5d, 0d, 0d);
+            spawnLoc.getWorld().spawn(spawnLoc, ArmorStand.class, armorStand -> {
+                armorStand.setCustomNameVisible(true);
+                armorStand.customName(Component.text("Testing: ")
+                    .append(Component.translatable("testplugin.armorstand.arg")));
+            });
+        });
     }
 }
```

</details>

With these changes, Components are now no longer translated recursively (as expected):
image

I'm not sure if recursive translation will be possible when saving the direct codec because of how recursive codecs work - or I've just messed something else up.

File: patches/server/0010-Adventure.patch
Comment (jpenilla) [inline patches/server/0010-Adventure.patch]: just use a ConcurrentHashMap instead

**Comment by jpenilla** [comment]
Just a heads up that you probably want to wait until #10028 and #10014 are merged to make more changes, as they touch the same areas.

File: patches/server/0010-Adventure.patch
Comment (jpenilla) [inline patches/server/0010-Adventure.patch]: I think this needs a null check for the Locale, falling back to the AdventureCodecs.COMPONENT_CODEC

File: patches/server/0010-Adventure.patch
Comment (jpenilla) [inline patches/server/0010-Adventure.patch]: I think once we convert to adventure all we need to do is pass the result into the relevant adventure Codec, like is done above. Then we don't need to make #translated public either.

File: patches/server/0010-Adventure.patch
Comment (jpenilla) [reply patches/server/0010-Adventure.patch]: Maybe we should just modify the #localizedCodec methods to check for null before using the Map and return the base/non-localized Codec instead

**Review by jpenilla — APPROVED** [review]

**Comment by jpenilla** [comment]
I think this misses some cases where ComponentSerialization.CODEC is used directly instead of passing the self codec, like show_text hover events. Solving that is more complicated though, so I think it's fine to leave for later and not block this improvement.

**Comment by booky10** [comment]
> This misses some cases where ComponentSerialization.CODEC is used directly instead of passing the self codec, like show_text hover events. Solving that is more complicated though, so I think it's fine to leave for later and not block this improvement.

Solving this properly would probably require to re-create every codec requiring the Component codec for every locale. A quick workaround for this would be to just pass everything (which has show_text hover events, etc.) through the adventure codec.

**Comment by jpenilla** [comment]
Yeah, both options there are not very pretty, which is why I think we should just leave it for when/if someone finds a use case that justifies it. The only thing I can think of would be translating datapack messages with a plugin, but that feels really niche, and if someone needs that to work they can open an issue.

**Pull request merged**
